core/docs/stardoc.bzl

```diff
@@ -0,0 +1,134 @@
+        deps: library dependencies of the `inputs` provided.
+        **kwargs: arguments for `stardoc`, most notably `symbol_names` to generate documentation for.
+    """
+
+    # necessary boilerplate: to build a Bazel library, we always need to include
+    # `bazel_tools` as a dependency.
+    # since we will combine toolchain libraries with the core library, but only
+    # need `bazel_tools` dependency once, only create if not already there.
+    if not native.existing_rule("bazel_tools"):
+        bzl_library(
+            name = "bazel_tools",
+            srcs = [
+                "@bazel_tools//tools:bzl_srcs",
+            ],
+        )
+
+    # massage first input file name into nice rule name
+    lib = Label(absolute_label(inputs[0])).name.split(".")[0]
+    bzl_library(
+        name = lib,
+        srcs = inputs,
+        visibility = ["//visibility:public"],
+        deps = [":bazel_tools"] + deps,
+    )
+
+    # generate documentation into transient file
+    out = "_{}".format(name)
+    stardoc("__{}".format(name), out, inputs[0], deps = [lib], **kwargs)
+
+    # create rule to copy documentation into source tree
+    # has to be run manually! set up a commit hook for convenience?
+    copy_files(
+        name = "update-{}".format(name),
+        data = [(out, name)],
+    )
+
+    if not error_message:
+        error_message = [
+            "{} is not up to date.",
+            "Please update it using the following command:",
+            "",
+            "bazel run //{}:update-{}",
+        ]
+        error_message = "\n".join(error_message).format(name, native.package_name(), name)
+
+    # create test that source tree is up to date with rendered documentation
+    compare_files(
+        name = "check-{}".format(name),
+        data = [(out, name)],
+        error_message = error_message,
+    )
+
+def stardoc(
+        name,
+        out,
+        input,
+        aspect_template = "@rules_nixpkgs_core//docs:templates/aspect.vm",
+        func_template = "@rules_nixpkgs_core//docs:templates/func.vm",
+        header_template = "@rules_nixpkgs_core//docs:templates/header.vm",
+        provider_template = "@rules_nixpkgs_core//docs:templates/provider.vm",
+        rule_template = "@rules_nixpkgs_core//docs:templates/rule.vm",
+        **kwargs):
+    _stardoc(
+        name = name,
+        out = out,
+        input = input,
+        aspect_template = aspect_template,
+        func_template = func_template,
+        header_template = header_template,
+        provider_template = provider_template,
+        rule_template = rule_template,
+        **kwargs
+    )
+
+def copy_files(name, data):
+    """copy list of files to workspace root"""
+    native.sh_binary(
+        name = name,
+        srcs = ["@rules_nixpkgs_core//docs:copy-files.sh"],
+        args = ["$(location {}) {}".format(a, b) for a, b in data],
+        data = [a for a, b in data],
+        env = {"PATH": "$(POSIX_CP)"},
```

Comment:
Does that really do the intended thing? I would expect this to set `PATH=/nix/store/.../bin/cp`, whereas I would expect us to want `PATH=/nix/store/.../bin`. Am I missing something?

Reply:
It did work as intended from a vanilla environment, but now that you say it, it surprises me, too. Will re-check.

Reply:
Yes, `$PATH` is indeed augmented with `/nix/store/.../bin/cp` and therefore ignored. The rest of `$PATH` still has a minimal `nix` environment including `coreutils`, but it seems not to come from `nixpkgs_git_repository`: changing the `nixpkgs` release there does not influence the store paths. That is weird, because I test this in
```
bash
unset NIX_PATH
unset NIX_PROFILES
unset NIX_PROFILE_DIR
export PATH=/nix/store/<...>-nix-2.3.16/bin:/usr/bin:/usr/sbin:/bin:/sbin
/nix/store/<...>-bazel-4.2.1/bin/bazel run //:update-README.md
```

I don't really see how one would actually take the proper `bin` path to `cp` without tinkering within the script. I would like to keep the script absolutely minimal and not concern itself with these environmental questions.

Reply:
It's defined in the Nix expression for Bazel itself here: https://github.com/NixOS/nixpkgs/blob/19d6e349b910b19d54b58a5b3af14ee0e1efedea/pkgs/development/tools/build-managers/bazel/bazel_4/default.nix#L78-L126